File: backend/internal/handlers/connection.go
package handlers

import (
	"encoding/json"
	"net/http"

	"ideal-ml/backend/internal/auth"
	"ideal-ml/backend/internal/config"
	"ideal-ml/backend/internal/github"
)

type ConnectionHandler struct {
	store  *config.Store
	client *github.Client
}

func NewConnectionHandler(store *config.Store, client *github.Client) *ConnectionHandler {
	return &ConnectionHandler{store: store, client: client}
}

type connectRequest struct {
	RepoOwner  string `json:"repoOwner"`
	RepoName   string `json:"repoName"`
	Branch     string `json:"branch"`
	ConfigPath string `json:"configPath"`
}

type connectionResponse struct {
	Status     string `json:"status"`
	ModelCount *int   `json:"modelCount,omitempty"`
	Error      string `json:"error,omitempty"`
	RepoOwner  string `json:"repoOwner,omitempty"`
	RepoName   string `json:"repoName,omitempty"`
	Branch     string `json:"branch,omitempty"`
	ConfigPath string `json:"configPath,omitempty"`
}

func (h *ConnectionHandler) Connect(w http.ResponseWriter, r *http.Request) {
	var req connectRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, connectionResponse{
			Status: "error",
			Error:  "invalid request body",
		})
		return
	}

	if req.RepoOwner == "" || req.RepoName == "" {
		writeJSON(w, http.StatusBadRequest, connectionResponse{
			Status: "error",
			Error:  "repoOwner and repoName are required",
		})
		return
	}

	cfg := &config.GitHubConfig{
		RepoOwner:  req.RepoOwner,
		RepoName:   req.RepoName,
		Branch:     req.Branch,
		ConfigPath: req.ConfigPath,
	}
	if cfg.Branch == "" {
		cfg.Branch = "main"
	}
	if cfg.ConfigPath == "" {
		cfg.ConfigPath = "models.yaml"
	}

	h.store.Set(cfg)

	token := auth.TokenFromContext(r.Context())
	count, err := h.client.TestConnection(token)
	if err != nil {
		h.store.Clear()
		writeJSON(w, http.StatusOK, connectionResponse{
			Status: "error",
			Error:  err.Error(),
		})
		return
	}

	writeJSON(w, http.StatusOK, connectionResponse{
		Status:     "connected",
		ModelCount: &count,
	})
}

func (h *ConnectionHandler) Disconnect(w http.ResponseWriter, r *http.Request) {
	h.store.Clear()
	h.client.InvalidateCache()
	writeJSON(w, http.StatusOK, connectionResponse{Status: "disconnected"})
}

func (h *ConnectionHandler) Status(w http.ResponseWriter, r *http.Request) {
	cfg := h.store.Get()
	if cfg == nil {
		writeJSON(w, http.StatusOK, connectionResponse{Status: "disconnected"})
		return
	}

	writeJSON(w, http.StatusOK, connectionResponse{
		Status:     "connected",
		RepoOwner:  cfg.RepoOwner,
		RepoName:   cfg.RepoName,
		Branch:     cfg.Branch,
		ConfigPath: cfg.ConfigPath,
	})
}

func writeJSON(w http.ResponseWriter, status int, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

File: backend/app/routes/connection.py
from fastapi import APIRouter, Depends, Request

from app.config import GitHubConfig

router = APIRouter()


def _require_auth():
    raise NotImplementedError("Must be overridden at app level")


@router.post("/api/connection")
async def connect(request: Request, token: str = Depends(_require_auth)):
    body = await request.json()

    repo_owner = body.get("repoOwner", "")
    repo_name = body.get("repoName", "")
    branch = body.get("branch", "") or "main"
    config_path = body.get("configPath", "") or "models.yaml"

    if not repo_owner or not repo_name:
        return {"status": "error", "error": "repoOwner and repoName are required"}

    cfg = GitHubConfig(
        repoOwner=repo_owner,
        repoName=repo_name,
        branch=branch,
        configPath=config_path,
    )

    store = request.app.state.config_store
    client = request.app.state.github_client

    store.set(cfg)

    try:
        count = await client.test_connection(token)
    except Exception as e:
        store.clear()
        return {"status": "error", "error": str(e)}

    return {"status": "connected", "modelCount": count}


@router.delete("/api/connection")
async def disconnect(request: Request, token: str = Depends(_require_auth)):
    request.app.state.config_store.clear()
    request.app.state.github_client.invalidate_cache()
    return {"status": "disconnected"}


@router.get("/api/connection/status")
async def status(request: Request, token: str = Depends(_require_auth)):
    cfg = request.app.state.config_store.get()
    if cfg is None:
        return {"status": "disconnected"}

    return {
        "status": "connected",
        "repoOwner": cfg.repo_owner,
        "repoName": cfg.repo_name,
        "branch": cfg.branch,
        "configPath": cfg.config_path,
    }
